## Option Pricing using Monte Carlo Simulation

Option pricing using Monte Carlo simulation and geometric brownian motion. 

Will calculate CALL and PUT Payoffs.

In [1]:
# Libraries

import numpy as np

In [2]:
S = 120                 # Spot price
K = 100                 # Strike price
T = 0.083               # Time to expiration (Years)
r = 0.07                # Discount rate (%)     
sigma = 0.2             # Volatility of stock(%)

Nsteps = 30             # Number of days option runs for
Nsimulations = 5        # Number of simulations 

In [3]:
# Brownian motion for daily prices:
dt = T / Nsteps

In [4]:
# Drift:
drift = (r-(sigma**2)/2) * dt

In [17]:
# Time:
a = sigma * np.sqrt(dt)

In [18]:
# Random variable: Creating random variables for each day of stock.

# Using standard normal distribution with mean=0, variance=1, (Nsimulation for Nsteps) 
x = np.random.normal(0,1,(Nsimulations, Nsteps))      
print(x)

[[-0.76620608  0.73013866 -0.29244232  0.45414429 -1.71942056 -0.47634849
   0.93882865  0.14413749  2.1681109  -1.12123506 -1.03963915  0.16523069
   1.6714749  -1.33839279  0.16029991 -1.06205689 -0.00295069 -1.09751654
  -0.21409834 -1.38540445 -1.01289809  1.76422255 -1.51121306 -0.92583176
   2.07637023  0.84017476  1.49052017  0.62679259 -0.41372466 -0.48210436]
 [-0.84850185 -0.22362511  1.18546634  2.07593008  1.92474663  0.48549559
   1.0105413  -0.58220529  0.14277337  1.11763883  2.03961298  0.55525826
  -0.1088304  -0.72943178  1.52374533  1.78526791  0.12083923  0.3193558
   1.46331531  1.01399219  0.526434   -0.55438656 -0.08961986 -1.47341491
   0.05240309 -0.35188193  0.39771989  1.59984473 -1.18977888 -0.1565989 ]
 [ 2.5665023   1.06732201  0.28008195 -0.19981052 -0.09940198 -0.85511381
  -0.06783743 -0.10220807 -1.05687085  0.70206692  0.6384195   0.9958273
  -0.435677    0.78658558  1.25152096  0.64183178 -0.46011892 -0.05094941
   0.20096224  0.77390271  0.94953508 

Above we can see that for each of the 5 arrays (simulations), there are 30 elements (steps). Each element represents one day.

In [7]:
# Matrix of stock price values

Smat = np.zeros((Nsimulations, Nsteps))     # Matrix initially loaded with zero values
Smat[:,0]+= S                               # put 120 (Spot price) on the first element of each of the 5 arrays
Smat

array([[120.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [120.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [120.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [120.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
       [120.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
          0.,   0.,   0.,   0., 

Above we can see that the each of the 5 arrays contain a value of 120 (Spot price) at the 1st element, with 0's afterwards. These 0's will be exchanged for a value based on the random variables generated prevoiusly.

In [8]:
# Brownian Motion: (from the 2nd element onwards, we have the Brownian motion values)
# Each simulation represents a different Brownian Motion path.

for i in range(1, Nsteps):
    Smat[:,i] += Smat[:,i-1] * np.exp(drift + a*x[:,i])
Smat

array([[120.        , 118.0521964 , 117.06167364, 114.08995088,
        114.97111187, 113.46191563, 115.12925944, 114.83310334,
        116.35980948, 118.04948379, 118.69566182, 119.62664285,
        120.75750939, 119.61398662, 120.86568137, 122.1121771 ,
        123.63489028, 124.53763013, 124.19505747, 123.16055699,
        123.07020073, 124.40227112, 123.49004274, 122.49752075,
        123.85470825, 126.62440178, 128.34427361, 130.61015903,
        129.47646403, 129.96907949],
       [120.        , 120.07521774, 119.93567375, 118.29690895,
        119.19376327, 119.80798554, 121.08854823, 118.4436794 ,
        116.6750918 , 118.48231515, 118.02579132, 115.74427433,
        117.52826697, 117.26645597, 116.00487051, 116.67045571,
        117.13711629, 117.55266061, 118.04972297, 117.17562517,
        117.5475024 , 119.1204332 , 116.29287415, 114.32194143,
        114.29900451, 112.74431127, 112.38794979, 111.66796154,
        110.98668815, 110.85098335],
       [120.        , 119.7217

Above we can see that we have 5 different Brownian Motion paths (simulations) for the stock price.

### Calulate Payoffs for CALL and PUT:

In [10]:
# CALL payoff values:

c = (Smat[:,-1]) - K        # Call = 'last value of Smat' subtract 'Strike Price'

for i in range(len(c)):
    if c[i] < 0:
        c[i] = 0            # where values are less than 1, make these values equal zero
    else:
        c[i] = c[i]         # otherwise take value as it is                
c

array([29.96907949, 10.85098335, 13.01844907, 40.86868074, 18.30096129])

In [11]:
# PUT payoff values:

p = K - (Smat[:,-1])        # Put = 'Strike Price' subtract the 'Spot price' 
for i in range(len(p)):
    if p[i] < 0:
        p[i] = 0            # where values are less than 1, make all values equal zero
    else:
        p[i] = p[i]         # otherwise take value as it is                
p

array([0., 0., 0., 0., 0.])

In [12]:
# Take the average of payoff simulations:

call_payoff = np.mean(c)
print('Call payoff =', call_payoff)

put_payoff = np.mean(p)
print('Put payoff =', put_payoff)

Call payoff = 22.601630789066302
Put payoff = 0.0


### Discount price back to present day:

In [15]:
call = call_payoff*np.exp(-r*T)
put = put_payoff*np.exp(-r*T)

In [16]:
print('Call payoff (Discounted to present day) =', call)
print('Put payoff (Discounted to present day) =', put)

Call payoff (Discounted to present day) = 22.470696047925152
Put payoff (Discounted to present day) = 0.0


We can see that the call payoff has value of 22.47, whereas the put payoff has a value of 0.0. This is because for these given spot and strike prices they only work in favor of a call option. Hence a Put option at these prices would not be executed. 